lib/graphs/digraph.nit

```diff
@@ -0,0 +1,606 @@
+	fun successors(u: V): Collection[V] is abstract
+
+	# Returns the vertices of this graph.
+	fun vertices: Collection[V] is abstract
+
+	# Returns the arcs of this graph.
+	fun arcs: Collection[Arc[V]] is abstract
+
+	## -------------------- ##
+	## Non abstract methods ##
+	## -------------------- ##
+
+	## ------------ ##
+	## Neighborhood ##
+	## ------------ ##
+
+	# Returns true if and only if `u` is a predecessor of `v`.
+	fun is_predecessor(u, v: V): Bool
+	do
+		return has_arc(u, v)
+	end
+
+	# Returns true if and only if `u` is a successor of `v`.
+	fun is_successor(u, v: V): Bool
+	do
+		return has_arc(v, u)
+	end
+
+	# Returns the incoming arcs of vertex `u`.
+	#
+	# If `u` is not in this graph, an empty array is returned.
+	fun incoming_arcs(u: V): Array[Arc[V]]
+	do
+		var arcs = new Array[Arc[V]]
+		for v in successors(u) do
+			arcs.add(new Arc[V](u, v))
+		end
+		return arcs
+	end
+
+	# Returns the outgoing arcs of vertex `u`.
+	#
+	# If `u` is not in this graph, an empty array is returned.
+	fun outgoing_arcs(u: V): Array[Arc[V]]
+	do
+		var arcs = new Array[Arc[V]]
+		for v in predecessors(u) do
+			arcs.add(new Arc[V](v, u))
+		end
+		return arcs
+	end
+
+	## ---------------------- ##
+	## String representations ##
+	## ---------------------- ##
+
+	redef fun to_s: String
+	do
+		var s = "vertices"
+		var t = "arcs"
+		if num_vertices <= 1 then s = "vertex"
+		if num_arcs <= 1 then t = "arc"
+		s = "Digraph of {num_vertices} {s} and {num_arcs} {t}\n"
+		s += "  Vertices: {vertices.join(" ")}\n"
+		s += "  Arcs: {arcs.join(" ")}"
+		return s
+	end
+
+	# Returns a GraphViz string representing this digraph.
+	fun to_graphviz: String
+	do
+		var s = "digraph \{\n"
+		# Writing the vertices
+		for u in vertices do
+			s += "  \"{u}\" [label=\"{u}\"];\n"
+		end
+		# Writing the arcs
+		for arc in arcs do
+			s += "  {arc.source} -> {arc.target};\n"
+		end
+		s += "\}"
+		return s
+	end
+
+	# ------- #
+	# Degrees #
+	# ------- #
+
+	# Returns the number of arcs whose target is `u`.
+	fun in_degree(u: V): Int
+	do
+		return predecessors(u).length
+	end
+
+	# Returns the number of arcs whose source is `u`.
+	fun out_degree(u: V): Int
+	do
+		return successors(u).length
+	end
+
+	# ------------------ #
+	# Paths and circuits #
+	# ------------------ #
+
+	# Returns true if and only if `vertices` is a path of this digraph.
+	fun is_path(vertices: SequenceRead[V]): Bool
+	do
+		for i in [0..vertices.length[ do
+			if not has_arc(vertices[i], vertices[i + 1]) then return false
+		end
+		return true
+	end
+
+	# Returns true if and only if `vertices` is a circuit of this digraph.
+	fun is_circuit(vertices: SequenceRead[V]): Bool
+	do
+		return vertices.is_empty or (is_path(vertices) and vertices.first == vertices.last)
+	end
+
+	# Returns a shortest path between vertices `u` and `v`.
+	#
+	# If no path exists between `u` and `v`, it returns `null`.
+	fun a_shortest_path(u, v: V): nullable List[V]
+	do
+		var queue = (new Array[V]).as_fifo
```

Comment:
from the doc:

~~~
# Note that some `Sequence`, like `Array`, have an inefficient `shift` implementation,
# thus are not the best candidates to serve as a FIFO.
~~~